src/promql/src/extension_plan/normalize.rs

```diff
@@ -0,0 +1,341 @@
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.input.schema()
+    }
+
+    fn output_partitioning(&self) -> Partitioning {
+        self.input.output_partitioning()
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        self.input.output_ordering()
+    }
+
+    fn maintains_input_order(&self) -> bool {
+        false
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.input.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::common::Result<Arc<dyn ExecutionPlan>> {
+        assert!(!children.is_empty());
+        Ok(Arc::new(Self {
+            offset: self.offset,
+            input: children[0].clone(),
+            metric: self.metric.clone(),
+        }))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<datafusion::execution::context::TaskContext>,
+    ) -> datafusion::common::Result<SendableRecordBatchStream> {
+        let baseline_metric = BaselineMetrics::new(&self.metric, partition);
+
+        let input = self.input.execute(partition, context)?;
+        Ok(Box::pin(SeriesNormalizeStream {
+            offset: self.offset,
+            schema: input.schema(),
+            input,
+            metric: baseline_metric,
+        }))
+    }
+
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default => {
+                write!(f, "PromSeriesNormalizeExec: offset=[{}]", self.offset)
+            }
+        }
+    }
+
+    fn metrics(&self) -> Option<MetricsSet> {
+        todo!()
+    }
+
+    fn statistics(&self) -> Statistics {
+        self.input.statistics()
+    }
+}
+
+pub struct SeriesNormalizeStream {
+    offset: Millisecond,
+
+    schema: SchemaRef,
+    input: SendableRecordBatchStream,
+    metric: BaselineMetrics,
+}
+
+impl SeriesNormalizeStream {
+    pub fn normalize(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
+        let ts_column_idx = self
+            .schema
+            .column_with_name(TIMESTAMP_COLUMN_NAME)
+            .expect("timestamp column not found")
+            .0;
+        // todo: maybe the input is not timestamp millisecond array
+        let ts_column = input
+            .column(ts_column_idx)
+            .as_any()
+            .downcast_ref::<PrimitiveArray<TimestampMillisecondType>>()
+            .unwrap();
+
+        // bias the timestamp column by offset
+        let ts_column_biased = Arc::new(TimestampMillisecondArray::from_iter(
+            ts_column.iter().map(|ts| ts.map(|ts| ts - self.offset)),
+        ));
+        let mut columns = input.columns().to_vec();
+        columns[ts_column_idx] = ts_column_biased;
+
+        // sort the record batch
+        let ordered_indices = compute::sort_to_indices(&columns[ts_column_idx], None, None)?;
```

Comment:
Could we assume the input is sorted? At least the storage engine sorts each series' data by time.

Reply:
There is a method `output_ordering` for plans to specify the order. But we set it to `None`. We can change it later with this place.